### ISSUE TITLE
Dismissing Command Window = 1s+ delay (New in 1.5.2 [401D])

### ISSUE THREAD

**Issue by ar-rm** [quicksilver/Quicksilver#2310]
Hi,

I just upgraded to [1.5.2 (401D)](https://github.com/quicksilver/Quicksilver/releases/tag/build401D) and immediately noticed a ~1s delay before the QS Command Window is dismissed.

IOW: Pressing `enter` to run a command, holding down the last letter of a search (to auto-run), pressing `esc` to leave the QS window, all leave the QS window on the screen for around 1sec. Then it disappears. It was instant Pre-1.5.2.

Worse: It's blocking action execution. i.e.: Launching an app now takes an extra 1s, because QS is getting hung up on its own command window.

These are logged (multiple times) to the `system.log` after each window dismissal:

```
Jan 19 12:59:44 WindowServer[225]: disable_update_timeout: UI updates were forcibly disabled by application "Quicksilver" for over 1.00 seconds. Server has re-enabled them.
Jan 19 12:59:45 WindowServer[225]: common_reenable_update: UI updates were finally reenabled by application "Quicksilver" after 1.50 seconds [0.67fps] (server forcibly re-enabled them after 1.00 seconds [1.00fps])
 ```

I did some issue searches and maybe this is related to:
#1687: "Quicksilver slow to "wake up" / "not responding"

I tried disabling the "Contacts" plugin (as suggested in that issue), running a `Force Catalog Rescan`, and restarting QS. It didn't help.

I'm tempted to downgrade to [401A](https://github.com/quicksilver/Quicksilver/releases/tag/build401A), but I'll wait for a little bit to see if anyone has any other suggestions and debugging ideas.

Thanks.

**Comment by ar-rm** [comment]
Possible cause?:

> I believe this error can be caused by taking too long inside of -drawRect: or some other part of the drawing machinery. It’s most likely a symptom, not the root problem—the root problem is your drawing is super-slow.

http://stackoverflow.com/a/21249196/886672

Also:
https://www.macissues.com/2015/01/08/how-to-manage-intermittent-lag-in-os-x-yosemite/

**Comment by skurfer** [comment]
In previous 1.5.x versions, there was a bug where the string you typed to search was being thrown out as soon as the window was dismissed (before it could be saved). Now, what you searched for and what you picked are saved before the window is dismissed. Maybe that’s taking longer than usual on your system for some reason.

**Comment by skurfer** [comment]
See if this is any better. https://cl.ly/0B282O2z3s0C

@pjrobertson I changed `-[QSMnemonics writeItems:]` to run on another thread asynchronously. Thoughts? I’m running the same build locally to see if there are any issues.

**Comment by ar-rm** [comment]
Thanks for the quick reply and build.

Unfortunately *no* change.

>Jan 20 11:10:52 WindowServer[225]: disable_update_timeout: UI updates were forcibly disabled by application "Quicksilver" for over 1.00 seconds. Server has re-enabled them.

> Jan 20 11:10:52 WindowServer[225]: common_reenable_update: UI updates were finally reenabled by application "Quicksilver" after 1.41 seconds [0.71fps] (server forcibly re-enabled them after 1.00 seconds [1.00fps])

I should also mention that it's the same if you try to dismiss the window via mouse press (outside of the window) without typing *anything*.

I also reverted to 401A (drag and drop, not preference swaps or anything) and it immediately cleared up.

**Comment by skurfer** [comment]
For future reference…

https://github.com/quicksilver/Quicksilver/compare/build401A...build401C

One or two changes to look over. 😐

**Comment by tiennou** [comment]
@ar-rm Are you Terminal-savvy ? I can give you a (convoluted) command that might help in pinpointing what happen...

**Comment by ar-rm** [comment]
@tiennou Yeah! Hit me.

**Comment by tiennou** [comment]
Use `sleep $delay; sample $pid $duration > ~/qs-sample`. Replace $pid with QS's pid, $delay with how many seconds you're comfortable triggering the bug (also, the sampling start is loong, you'll see).$duration is how many seconds of sampling to do (around 3-4 maybe, IIRC the default is 10).

You'll get a "standard" Activity Monitor file in ~. If you can time it right (the sample start *right before* the lag and goes on through), it might help to pinpoint what happens. HTH !

**Comment by pjrobertson** [comment]
@skurfer - I doubt it's to do with saving mnemonics. If that's taking several seconds then something's seriously wrong.

**Comment by ar-rm** [comment]
@tiennou Ok, I think I got it.

There was a short delay, the Sample process started, I recreated the issue (QS window, then `esc` key = delay), the Sample finished and wrote to the destination file:

[Quicksilver_2017-01-22_230358_DDCr.sample.txt](https://github.com/quicksilver/Quicksilver/files/722598/Quicksilver_2017-01-22_230358_DDCr.sample.txt)

I hope this helps.

**Comment by tiennou** [comment]
`+[QSObject purgeImagesAndChildrenOlderThan:]` 🔥 

Seems like the set-wrangling we do here came back to haunt us...

**Comment by skurfer** [comment]
Strange. Most of those changes were from the 1.2.x era, so not even close to new. You think it’s copying the sets in a huge catalog that could explain the delay?

I also see that we’re using concurrent enumeration there. This is probably another one of those places where there aren’t enough objects for us to benefit from it.

**Comment by tiennou** [comment]
I suspect Apple changed something in Sierra w.r.t to UTI conformance, which makes our current `-[QSObject isEqual:]` slower. Under that `purgeImage` tree-sample, we're calling `QSUTIForAnyTypeString` (at least) two times (-count/-objectForType:) in a row, which now shows up as a delay because that purge set might be big enough when the interface closes.

**Comment by skurfer** [comment]
I don’t think we’re caching the result of these UTI lookups anywhere. Maybe it’s time we start?

**Comment by skurfer** [comment]
Give this test build a try and let me know if it’s any better. The version should show as “1.5.2:mine”.

https://cl.ly/0c3v0M031t3L

**Comment by ar-rm** [comment]
Unfortunately this version didn't even fully launch.

- It didn't crash.
- But I also can't initiate it either. Nothing happens after I press the keyboard shortcut.
- Nothing seems to have been written to system.log.
- It does show up as running in Activity Monitor.
- But I have to force quit it to shut it down.
- I took a sample, via AM, while it was hanging: [Sample of Quicksilver - 1.5.2-mine.txt](https://github.com/quicksilver/Quicksilver/files/756547/Sample.of.Quicksilver.-.1.5.2-mine.txt)
- I'm still able to launch 1.4.2.

**Comment by skurfer** [comment]
Sorry. A debugging setting snuck into the release build. See if this is any better.

https://cl.ly/420r2R0z3R3d

**Comment by ar-rm** [comment]
- Good News: QS works.
- Bad News: The original issue is still there.

However this time, *nothing* is logged to the `system.log` while QS is hanging.

**Comment by skurfer** [comment]
Give this a try: https://cl.ly/011O462y3r3C

**Comment by ar-rm** [comment]
A Ha! You got it!

*1.5.2:mine (401D) - 2017-02-28]*

- No delays when calling or dismissing the QS window
- No messages logged to the Console (system.log)

Is this built "safe" to use, or should I wait for an official beta or release version?

Thanks!

**Comment by skurfer** [comment]
It’s a proof of concept (which appears to be proven) and not *quite* right for an official release, but should be safe enough.

If anything is going to go wrong, it *should* be confined to Quicksilver’s catalog and not affect anything else on your machine. (You might see some items duplicated, for example.)

**Comment by ar-rm** [comment]
> It’s a proof of concept (which appears to be proven)

What was it? (Without taking too much time)

> not quite right for an official release, but should be safe enough.

Ok. I'll stick with 1.4.2 for now, and upgrade to < 1.5.2 when it's out.

Thanks again!

**Comment by skurfer** [comment]
> What was it? (Without taking too much time)

Basically, the code that compares two items in the catalog (to avoid duplicates, etc) was taking a lot longer than it should have, and getting called *way, way* more often than we realized. This was especially noticeable when you launched QS because it reloaded the cached catalog from disk on the main thread (meaning the interface can’t do anything until it finishes). Normal catalog scanning happens in the background, so even if it takes a while, it doesn’t freeze the app.

Here’s another build with additional improvements that should be more or less what ends up in the next release. Feel free to try it, or stay on 1.4.2.

https://cl.ly/1p1K3I2e1F3M

**Comment by ar-rm** [comment]
Thanks for that information and the hard work to fix it *and* in general :-)

I just updated to 1.5.3 (401E) [Release]

Still fixed.
Thanks!

One related, but probably different thing I've noticed:
- If *feels* like there's a very slight delay between pressing "return" and QS doing whatever was chosen (ex: Opening an app).
- It's almost like QS is waiting for keyUP instead of keyDOWN to *go*.
- I don't know if this is new or been around for a while, or forever. I'd say it feels newish because QS has always been *so* fast. As a result, *any* slowdown throws up a red flag.

**Comment by skurfer** [comment]
> If *feels* like there's a very slight delay between pressing "return" and QS doing whatever was chosen (ex: Opening an app).

A co-worker of mine sees a very noticeable delay when using “Paste as Plain Text”, butI’ve ben unable to reproduce it. A lot of changes have been made since I last asked him, though.

**Comment by pjrobertson** [comment]
@skurfer - did this ever get merged into master? This looks like a useful thing to have, and I know we call the `UTI` stuff way too much.

**Comment by pjrobertson** [comment]
Yeah, it looks like this was merged in #2312 and fixed in 1.5.3. Great!
